### PR TITLE
[Radar] Add BGP anomaly detection examples

### PR DESCRIPTION
**Pull request by digizeph**
## Overview

This PR adds BGP anomaly detection examples and fixes a couple small issues under Radar.

## Changes

1. add `BGP anomalies` page under `radar/investigate` section
2. remove Radar index page beta tag
3. remove one duplicate line in DNS examples page

### CODE REVIEW

**Review by G4brym — APPROVED** [review]

**Review by celso — APPROVED** [review]

File: content/radar/investigate/bgp-anomalies.md [hunk 1]
```diff
@@ -0,0 +1,337 @@
+---
+pcx_content_type: reference
+title: BGP anomalies
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
title: BGP anomalies (beta)
```

File: content/radar/investigate/bgp-anomalies.md [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+weight: 3
+---
+
+{{<beta>}}BGP anomalies{{</beta>}}
+
+Access Cloudflare Radar BGP Anomaly Detection results.
+
+You will need to first create an API token that has `User:User Details` permission included. All the following examples
+should work with a free-tier Cloudflare account.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
To access Cloudflare Radar BGP Anomaly Detection results, you will first need to create an API token that includes a `User:User Details` permission. All the following examples should work with a free-tier Cloudflare account.
```

Comment (marciocloudflare) [reply content/radar/investigate/bgp-anomalies.md]: Active voice is always preferred to passive :)

File: content/radar/investigate/bgp-anomalies.md [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+## Search BGP Hijack Events
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]: Unless this is a feature name, it should be lowecase

File: content/radar/investigate/bgp-anomalies.md [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+- `peer_asns`: the AS numbers of the route collector peers who observed the hijack messages
+- `prefixes`: the affected prefixes
+- `hijacker_asn` and `victim_asns`: the potential hijacker ASN and victim ASNs
+- `confidence_score`: a quantitative score describing how confident the system is for this event being a hijack
+  - 1-3: low confidence
+  - 4-7: medium confidence
+  - 8-above: high confidence
+- `tags`: the evidence collected for the events, each is also associated with a score that affects the overall confidence score
+  - a positive score indicates that the event is *more likely* to be a hijack
+  - a negative score indicates that the event is *less likely* to be a hijack
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
In the response we can learn about the following information about each event:
- `hijack_msg_count`: the number of potential BGP hijack messages observed from all peers.
- `peer_asns`: the AS numbers of the route collector peers who observed the hijack messages.
- `prefixes`: the affected prefixes.
- `hijacker_asn` and `victim_asns`: the potential hijacker ASN and victim ASNs.
- `confidence_score`: a quantitative score describing how confident the system is for this event being a hijack:
  - 1-3: low confidence.
  - 4-7: medium confidence.
  - 8-above: high confidence.
- `tags`: the evidence collected for the events. Each `tag` is also associated with a score that affects the overall confidence score:
  - a positive score indicates that the event is *more likely* to be a hijack.
  - a negative score indicates that the event is *less likely* to be a hijack.
```

File: content/radar/investigate/bgp-anomalies.md [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+Users can further filter out low-confidence events by attaching a `minConfidence=8` parameter, which will return only the events with `confidence_score` to be 8 our higher.
+```bash
+curl -X GET "https://api.cloudflare.com/client/v4/radar/bgp/hijacks/events?invlovedAsn=64512&format=json&per_page=10&minConfidence=8" \
+     -H "Authorization: Bearer <API_TOKEN>"
+```
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
Users can further filter out low-confidence events by attaching a `minConfidence=8` parameter, which will return only events with a `confidence_score` of `8` or higher.

```bash
curl -X GET "https://api.cloudflare.com/client/v4/radar/bgp/hijacks/events?invlovedAsn=64512&format=json&per_page=10&minConfidence=8" \
     -H "Authorization: Bearer <API_TOKEN>"
```
```

File: content/radar/investigate/bgp-anomalies.md [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+
+## Search BGP Route Leak Events
+
+BGP route leak is another type of BGP anomalies that Cloudflare Radar detects. Currently, we focus on detecting specifcally
+the `provider-customer-provider` type of route leak. You can learn more about our design and methodology in [our blog post][route-leak-blog-post].
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
BGP route leak is another type of BGP anomaly that Cloudflare Radar detects. Currently, we focus on detecting specifcally the `provider-customer-provider` type of route leak. You can learn more about our design and methodology in [our blog post][route-leak-blog-post].
```

File: content/radar/investigate/bgp-anomalies.md [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+}
+```
+
+In the response we can learn about the following information about each event:
+- `leak_asn`: the AS who potentially caused the leak
+- `leak_seg`: the AS path segment observed and believed to be a leak
+- `min_ts` and `max_ts`: the earliest and latest timestamps of the leak announcements
+- `leak_count`: the total number of BGP route leak announcements observed
+- `peer_count`: the number of route collector peers observed the leak
+- `prefix_count` and `origin_count`: the number of prefixes and origin ASes affected by the leak
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
In the response we can learn about the following information about each event:
- `leak_asn`: the AS who potentially caused the leak.
- `leak_seg`: the AS path segment observed and believed to be a leak.
- `min_ts` and `max_ts`: the earliest and latest timestamps of the leak announcements.
- `leak_count`: the total number of BGP route leak announcements observed.
- `peer_count`: the number of route collector peers observed the leak.
- `prefix_count` and `origin_count`: the number of prefixes and origin ASes affected by the leak.
```

File: content/radar/investigate/bgp-anomalies.md [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+
+## Send Alerts for BGP Hijacks
+
+In this example, we will show you how you can build a simple Cloudflare Workers app that sends out alerts for BGP hijacks
+relevant to a given ASN using webhook (works for Google Hangouts, Discord, Telegram, etc) or Emails.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
In this example, we will show you how you can build a simple Cloudflare Workers app that sends out alerts for BGP hijacks relevant to a given ASN using webhooks (works for Google Hangouts, Discord, Telegram, etc) or email.
```

File: content/radar/investigate/bgp-anomalies.md [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+
+We will use Cloudflare Workers as the platform and use its Cron Triggers to periodically check for new alerts.
+
+For the app, we would like it to do the following things:
+1. fetch from Cloudflare API with a given API token
+2. check against Cloudflare KV to know what events are new
+3. construct messages for new hijacks and send out alerts via webhook triggers
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
- Fetch from Cloudflare API with a given API token.
- Check against Cloudflare KV to know what events are new.
- Construct messages for new hijacks and send out alerts via webhook triggers.
```

File: content/radar/investigate/bgp-anomalies.md [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+
+### Worker App Setup
+
+We will start with setting up a Cloudflare Worker app using `wrangler`.
+
+Let's first create a new workers app in a local directory
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
First, create a new Workers app in a local directory:

```

File: content/radar/investigate/bgp-anomalies.md [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+```bash
+wrangler init hijack-alerts
+```
+When prompt to select a type of worker, choose ` Scheduled handler`.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion

When prompted to select a type of Worker, choose ` Scheduled handler`.
```

File: content/radar/investigate/bgp-anomalies.md [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+
+In your `wrangler.toml` file, change the default checking frequency (once per hour) to what you like. Here is an example
+of configuring the workers to run the script five minutes.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md:184]:
```suggestion
of configuring the workers to run the script five minutes.

```

File: content/radar/investigate/bgp-anomalies.md [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+name = "hijack-alerts"
+main = "src/index.js"
+compatibility_date = "2023-04-27"
+
+[triggers]
+crons = [ "*/5 * * * *" ]
+```
+
+In this example, we will also need to use Cloudflare KV to save the latest checked event IDs which allows us to know what events are new.
+Once you have created a KV (it's free!), you can head back to the `wranglers.toml` file and add the following sections in.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
In this example, we will also need to use Cloudflare KV to save the latest checked event IDs which allows us to know what events are new. Once you have created a KV, you can head back to the `wranglers.toml` file and add the following sections:

```

File: content/radar/investigate/bgp-anomalies.md [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+```toml
+[[kv_namespaces]]
+binding = "HIJACKS_KV"
+id = "KV_ID_FOR_PRODUCTION"
+preview_id = "TEMPORARY_KV_FOR_DEV_ENVIRONMENT"
+```
+
+### Fetch for newly detected BGP hijacks
+
+Let's start with the API fetching function.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
Start with the API fetching function.
```

File: content/radar/investigate/bgp-anomalies.md [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+
+The following `apiFetch(env, paramsStr)` handles taking in a request parameters string, construct proper headers and
+fetch from the Cloudflare API BGP hijacks endpoint.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md:209]:
```suggestion
fetch from the Cloudflare API BGP hijacks endpoint.

```

File: content/radar/investigate/bgp-anomalies.md [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+
+  if(!res.ok){
+    console.log(JSON.stringify(res))
+    return null
+  }
+  return await (res).json()
+}
+```
+The `env` parameter is passed in from the caller and we do not need to worry about construct it. The `paramsStr` is a
+string variable that holds the query parameters in a query URL.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion

The `env` parameter is passed in from the caller and we do not need to worry about construct it. The `paramsStr` is a string variable that holds the query parameters in a query URL.
```

File: content/radar/investigate/bgp-anomalies.md [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+
+Now in our main cron trigger function, we will need to construct the query parameters and call the API fetch function.
+The default cron trigger worker script is defined as the follows:
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md:231]:
```suggestion
The default cron trigger worker script is defined as the follows:

```

File: content/radar/investigate/bgp-anomalies.md [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+```javascript
+
+export default {
+    async scheduled(controller, env, ctx) {
+    ...
+    }
+}
+```
+In our example, we use the `env` variables to get the runtime variables like the TOKEN and ASN of interest, and Cloudflare
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md:242]:
```suggestion

In our example, we use the `env` variables to get the runtime variables like the TOKEN and ASN of interest, and Cloudflare
```

File: content/radar/investigate/bgp-anomalies.md [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+KV bindings. We do not use of the `contorller` and `ctx` variables in this example.
+
+First, we will need to learn about what are the new events. We define new events as the events the app has not yet processed.
+We use the Cloudflare KV bucket previously created and defined (`HIJACKS_KV`) to save and retrieve the most recent
+processed event ID.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md:247]:
```suggestion
processed event ID.

```

File: content/radar/investigate/bgp-anomalies.md [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+```javascript
+let kv_latest_id = parseInt(await env.HIJACKS_KV.get("latest_id"));
+const first_batch = isNaN(kv_latest_id);
+```
+
+The main loop that checks for the most recent events looks like this. (Some of the validation code is skipped.)
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
The main loop that checks for the most recent events looks like this (some of the validation code is skipped):

```

File: content/radar/investigate/bgp-anomalies.md [hunk 21]
```diff
@@ -0,0 +1,337 @@
...
+    }
+    if(reached_last){
+      break
+    }
+    page += 1;
+}
+```
+
+
+Now that we have all the newly detected events saved in `new_events` variable, we can then send out alerts:
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md:290]:
```suggestion
Now that we have all the newly detected events saved in `new_events` variable, we can then send out alerts:

```

File: content/radar/investigate/bgp-anomalies.md [hunk 22]
```diff
@@ -0,0 +1,337 @@
...
+	await fetch(webhook_url, {
+		method: 'POST',
+		headers: {
+			'Content-Type': 'application/json; charset=UTF-8',
+		},
+		body: data,
+	});
+}
+```
+Note that the webhook is considered secret and should be set to the environment via `wrangler secret put WEBHOOK_URL` command.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md:335]:
```suggestion

Note that the webhook is considered secret and should be set to the environment via `wrangler secret put WEBHOOK_URL` command.
```

File: content/radar/investigate/bgp-anomalies.md [hunk 23]
```diff
@@ -0,0 +1,337 @@
...
+
+The last step is to publish the app with command `wrangler publish` and the app should be up and running on your Cloudflare
+account, and will be triggered to execute every 5 minutes.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
account, and will be triggered to execute every five minutes.
```

File: content/radar/investigate/bgp-anomalies.md [hunk 24]
```diff
@@ -0,0 +1,337 @@
...
+
+## Next steps
+
+Please check out our [developer portal][api-portal] for detailed API documentation.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
Refer to our [API documentationl][api-portal] for more information on these topics.
```

File: content/radar/investigate/bgp-anomalies.md [hunk 25]
```diff
@@ -0,0 +1,337 @@
...
+
+[api-portal]: https://developers.cloudflare.com/api
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]: This should link directly to Radar's API

**Review by marciocloudflare — CHANGES_REQUESTED** [review]

File: content/radar/investigate/bgp-anomalies.md (hunk 25 above)
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
[api-portal]: https://developers.cloudflare.com/api/
```

**Review by marciocloudflare — CHANGES_REQUESTED** [review]

File: content/radar/investigate/bgp-anomalies.md (hunk 25 above)
Comment (digizeph) [reply content/radar/investigate/bgp-anomalies.md]: Updated to radar api doc in the new commit

File: content/radar/investigate/bgp-anomalies.md (hunk 3 above)
Comment (digizeph) [reply content/radar/investigate/bgp-anomalies.md]: Changed all section titles to have upper case only for first letter

File: content/radar/investigate/bgp-anomalies.md [hunk 26]
```diff
@@ -0,0 +1,391 @@
...
+
+The last step is to publish the app with command `wrangler publish` and the app should be up and running on your Cloudflare
+account, and will be triggered to execute every five minutes.
+
+### Send email alerts from Workers
+
+If your app's domain has enabled [Email Routing][email-routing] feature, you can also send email alerts directly from
+your Workers app. You can learn more about the email-sending feature following [this tutorial][email-workers-tutorial].
+
+For this alerter app, you will need to add the `Emails binding` to the [`wrangler.toml`][wrangler-send-email] file.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
If you have [Email Routing][email-routing] enabled for your domain, you can also send email alerts directly from Workers. Refer to [Send emails from Workers][email-workers-tutorial] to learn more.

For this alert to work, you will need to configure the proper email bindings in the [`wrangler.toml`][wrangler-send-email] file.
```

File: content/radar/investigate/bgp-anomalies.md [hunk 27]
```diff
@@ -0,0 +1,391 @@
...
+
+```toml
+send_email = [
+    {type = "send_email", name = "SEND_EMAIL_BINDING", destination_address = "<YOUR_EMAIL>@example.com"},
+]
+```
+
+Then you can create an email sending function like the follows to send alert emails to your configured destination address.
```
Comment (marciocloudflare) [inline content/radar/investigate/bgp-anomalies.md]:
```suggestion
Then, you can create an email-sending function to send alert emails to your configured destination address:

```

**Review by marciocloudflare — CHANGES_REQUESTED** [review]

**Review by marciocloudflare — APPROVED** [review]

**Pull request merged**